<p style="font-family:'Tw Cen MT';font-size:14px"> P100 - Final Project - Part III</p>

# SQL Assignment
<p style="font-family:'Tw Cen MT';font-size:16px">by Isaac Nahmias, April 2022</p>

***

<h1>Table of Contents<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#Intoduction" data-toc-modified-id="Intoduction-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>Intoduction</a></span><ul class="toc-item"><li><span><a href="#Abstract" data-toc-modified-id="Abstract-1.1"><span class="toc-item-num">1.1&nbsp;&nbsp;</span>Abstract</a></span></li><li><span><a href="#Analysis-Objective" data-toc-modified-id="Analysis-Objective-1.2"><span class="toc-item-num">1.2&nbsp;&nbsp;</span>Analysis Objective</a></span></li><li><span><a href="#Importing-the-tables-and-relevant-libaries" data-toc-modified-id="Importing-the-tables-and-relevant-libaries-1.3"><span class="toc-item-num">1.3&nbsp;&nbsp;</span>Importing the tables and relevant libaries</a></span></li><li><span><a href="#Data-Description" data-toc-modified-id="Data-Description-1.4"><span class="toc-item-num">1.4&nbsp;&nbsp;</span>Data Description</a></span><ul class="toc-item"><li><span><a href="#Books-Table:" data-toc-modified-id="Books-Table:-1.4.1"><span class="toc-item-num">1.4.1&nbsp;&nbsp;</span>Books Table:</a></span></li><li><span><a href="#Authors-Table" data-toc-modified-id="Authors-Table-1.4.2"><span class="toc-item-num">1.4.2&nbsp;&nbsp;</span>Authors Table</a></span></li><li><span><a href="#Publishers-Table" data-toc-modified-id="Publishers-Table-1.4.3"><span class="toc-item-num">1.4.3&nbsp;&nbsp;</span>Publishers Table</a></span></li><li><span><a href="#Ratings-Table" data-toc-modified-id="Ratings-Table-1.4.4"><span class="toc-item-num">1.4.4&nbsp;&nbsp;</span>Ratings Table</a></span></li><li><span><a href="#Reviews-Table" data-toc-modified-id="Reviews-Table-1.4.5"><span class="toc-item-num">1.4.5&nbsp;&nbsp;</span>Reviews Table</a></span></li></ul></li></ul></li><li><span><a href="#Tasks" data-toc-modified-id="Tasks-2"><span class="toc-item-num">2&nbsp;&nbsp;</span>Tasks</a></span><ul class="toc-item"><li><span><a href="#Find-the-number-of-books-released-after-January-1,-2000:" data-toc-modified-id="Find-the-number-of-books-released-after-January-1,-2000:-2.1"><span class="toc-item-num">2.1&nbsp;&nbsp;</span>Find the number of books released after January 1, 2000:</a></span></li><li><span><a href="#Find-the-number-of-user-reviews-and-the-average-rating-for-each-book:" data-toc-modified-id="Find-the-number-of-user-reviews-and-the-average-rating-for-each-book:-2.2"><span class="toc-item-num">2.2&nbsp;&nbsp;</span>Find the number of user reviews and the average rating for each book:</a></span></li><li><span><a href="#Identify-the-publisher-that-has-released-the-greatest-number-of-books-with-more-than-50-pages:" data-toc-modified-id="Identify-the-publisher-that-has-released-the-greatest-number-of-books-with-more-than-50-pages:-2.3"><span class="toc-item-num">2.3&nbsp;&nbsp;</span>Identify the publisher that has released the greatest number of books with more than 50 pages:</a></span></li><li><span><a href="#Identify-the-author-with-the-highest-average-book-rating:" data-toc-modified-id="Identify-the-author-with-the-highest-average-book-rating:-2.4"><span class="toc-item-num">2.4&nbsp;&nbsp;</span>Identify the author with the highest average book rating:</a></span></li><li><span><a href="#Find-the-average-number-of-text-reviews-among-users-who-rated-more-than-50-books:" data-toc-modified-id="Find-the-average-number-of-text-reviews-among-users-who-rated-more-than-50-books:-2.5"><span class="toc-item-num">2.5&nbsp;&nbsp;</span>Find the average number of text reviews among users who rated more than 50 books:</a></span></li></ul></li><li><span><a href="#Project-Sammary:" data-toc-modified-id="Project-Sammary:-3"><span class="toc-item-num">3&nbsp;&nbsp;</span>Project Sammary:</a></span></li></ul></div>

## Intoduction

### Abstract

The coronavirus took the entire world by surprise, changing everyone's daily routine. City dwellers no longer spent their free time outside, going to cafes and malls; more people were home, reading books. That attracted the attention of startups that rushed to develop new apps for book lovers.

I've been given a database of one of the services competing in this market. It contains data on books, publishers, authors, and customer ratings and reviews of books.

### Analysis Objective

This information will be used to generate a value proposition for a new product.

### Importing the tables and relevant libaries

In [1]:
import pandas as pd
from sqlalchemy import create_engine

In [2]:
db_config = {'user': ***,         
             'pwd': ***,             
             'host': ***,
             'port': ***,                        
             'db': ***,}          

connection_string = 'postgresql://{}:{}@{}:{}/{}'.format(db_config['user'],
                                                         db_config['pwd'],
                                                         db_config['host'],
                                                         db_config['port'],
                                                         db_config['db'])

engine = create_engine(connection_string, connect_args={'sslmode':'require'})

### Data Description

#### Books Table:
Contains data on books:
* `book_id`
* `author_id`
* `title`
* `num_pages` — number of pages
* `publication_date`
* `publisher_id`


In [3]:
pd.io.sql.read_sql("SELECT * FROM books LIMIT 5", con = engine)

book_id  author_id                                              title  \
0        1        546                                       'Salem's Lot   
1        2        465                 1 000 Places to See Before You Die   
2        3        407  13 Little Blue Envelopes (Little Blue Envelope...   
3        4         82  1491: New Revelations of the Americas Before C...   
4        5        125                                               1776   

   num_pages publication_date  publisher_id  
0        594       2005-11-01            93  
1        992       2003-05-22           336  
2        322       2010-12-21           135  
3        541       2006-10-10           309  
4        386       2006-07-04           268

#### Authors Table

Contains data on authors:
* `author_id`
* `author`


In [4]:
pd.io.sql.read_sql("SELECT * FROM authors LIMIT 5", con = engine)

author_id                          author
0          1                      A.S. Byatt
1          2  Aesop/Laura Harris/Laura Gibbs
2          3                 Agatha Christie
3          4                   Alan Brennert
4          5        Alan Moore/David   Lloyd

#### Publishers Table

Contains data on publishers:
* `publisher_id`
* `publisher`


In [5]:
pd.io.sql.read_sql("SELECT * FROM publishers LIMIT 5", con = engine)

publisher_id                          publisher
0             1                                Ace
1             2                           Ace Book
2             3                          Ace Books
3             4                      Ace Hardcover
4             5  Addison Wesley Publishing Company

#### Ratings Table

Contains data on user ratings:
* `rating_id`
* `book_id`
* `username` — the name of the user who rated the book
* `rating`

In [6]:
pd.io.sql.read_sql("SELECT * FROM ratings LIMIT 5", con = engine)

rating_id  book_id       username  rating
0          1        1     ryanfranco       4
1          2        1  grantpatricia       2
2          3        1   brandtandrea       5
3          4        2       lorichen       3
4          5        2    mariokeller       2

#### Reviews Table

Contains data on customer reviews:
* `review_id`
* `book_id`
* `username` — the name of the user who reviewed the book
* `text` — the text of the review

In [7]:
pd.io.sql.read_sql("SELECT * FROM reviews LIMIT 5", con = engine)

review_id  book_id       username  \
0          1        1   brandtandrea   
1          2        1     ryanfranco   
2          3        2       lorichen   
3          4        3  johnsonamanda   
4          5        3    scotttamara   

                                                text  
0  Mention society tell send professor analysis. ...  
1  Foot glass pretty audience hit themselves. Amo...  
2  Listen treat keep worry. Miss husband tax but ...  
3  Finally month interesting blue could nature cu...  
4  Nation purpose heavy give wait song will. List...

***
## Tasks

### Find the number of books released after January 1, 2000:

In [8]:
sql = """
SELECT
    COUNT(book_id) AS books
FROM
    books
WHERE
    CAST(publication_date AS date) > '2000-01-01'
"""

pd.read_sql_query(sql, engine)

books
0    819

> **819** books has been released after Jan 1, 2000.

### Find the number of user reviews and the average rating for each book:

In [9]:
sql = """
SELECT 
    b.title,
    COUNT(DISTINCT re.review_id) AS reviews,
    AVG(ra.rating) AS avg_rating
FROM
    books AS b
LEFT OUTER JOIN ratings AS ra
    ON b.book_id = ra.book_id
LEFT OUTER JOIN reviews AS re
    ON b.book_id = re.book_id
GROUP BY b.book_id
ORDER BY COUNT(DISTINCT re.review_id) DESC
"""

pd.read_sql_query(sql, engine)

title  reviews  avg_rating
0                              Twilight (Twilight  #1)        7    3.662500
1                                  Water for Elephants        6    3.977273
2                                     The Glass Castle        6    4.206897
3    Harry Potter and the Prisoner of Azkaban (Harr...        6    4.414634
4    The Curious Incident of the Dog in the Night-Time        6    4.081081
..                                                 ...      ...         ...
995    Anne Rice's The Vampire Lestat: A Graphic Novel        0    3.666667
996                            The Natural Way to Draw        0    3.000000
997   The Cat in the Hat and Other Dr. Seuss Favorites        0    5.000000
998                          Essential Tales and Poems        0    4.000000
999  Disney's Beauty and the Beast (A Little Golden...        0    4.000000

[1000 rows x 3 columns]

### Identify the publisher that has released the greatest number of books with more than 50 pages:

In [11]:
sql = """
SELECT 
    b.publisher_id,
    p.publisher,
    COUNT(DISTINCT b.book_id) AS books
FROM books AS b
INNER JOIN publishers AS p
    ON b.publisher_id = p.publisher_id
WHERE b.num_pages > 50
GROUP BY b.publisher_id, p.publisher
ORDER BY COUNT(DISTINCT b.book_id) DESC
LIMIT 1
"""

pd.read_sql_query(sql, engine)

publisher_id      publisher  books
0           212  Penguin Books     42

> ***Penguin Books*** is the publisher that has released the greatest number of books with more than 50 pages.

### Identify the author with the highest average book rating:

look only at books with at least 50 ratings

In [14]:
sql = """
SELECT
    book_ratings.author_id, a.author, AVG(book_ratings.avg_rating) AS avg_rating
FROM (
    SELECT
        b.book_id,
        b.author_id,
        COUNT(ra.rating_id) AS ratings,
        AVG(ra.rating) AS avg_rating
    FROM
        books AS b
    INNER JOIN ratings AS ra
        ON b.book_id = ra.book_id
    GROUP BY b.book_id
    HAVING
        COUNT(ra.rating_id) >= 50
    ) AS book_ratings
INNER JOIN authors AS a
    ON book_ratings.author_id = a.author_id
GROUP BY book_ratings.author_id, a.author
ORDER BY AVG(book_ratings.avg_rating) DESC
LIMIT 1
"""

pd.read_sql_query(sql, engine)

author_id                      author  avg_rating
0        236  J.K. Rowling/Mary GrandPré    4.283844

> The author with the highest average book rating is **J.K. Rowling** (a.k.a. Mary GrandPré).

### Find the average number of text reviews among users who rated more than 50 books:

In [13]:
sql = """
SELECT AVG(t.reviews)
FROM (
    SELECT
        ra.username,
        COUNT(DISTINCT ra.book_id) AS ratings,
        COUNT(DISTINCT re.review_id) AS reviews
    FROM
        ratings AS ra
    INNER JOIN reviews AS re
        ON ra.username = re.username
    GROUP BY ra.username
    HAVING
        COUNT(DISTINCT ra.book_id) > 50
    ) AS t
"""

pd.read_sql_query(sql, engine)

avg
0  24.333333

 > The average number of text reviews among users who rated more than 50 books is about **24**.

***
## Project Sammary:

- Find the number of books released after January 1, 2000:
> 819 books has been released after Jan 1, 2000.
- Find the number of user reviews and the average rating for each book:
> *Twilight* is the book with the highest number of reviews - 7, and an average rating of 3.7.
- Identify the publisher that has released the greatest number of books with more than 50 pages:
> Penguin Books is the publisher that has released the greatest number of books (42 books) with more than 50 pages.
- Identify the author with the highest average book rating (look only at books with at least 50 ratings):
> The author with the highest average book rating is J.K. Rowling (a.k.a. Mary GrandPré), with an average books rating of 4.28.
- Find the average number of text reviews among users who rated more than 50 books:
> The average number of text reviews among users who rated more than 50 books is about 24.